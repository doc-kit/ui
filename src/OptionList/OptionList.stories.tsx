import { array, boolean, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import * as React from 'react';
import { withReadme } from 'storybook-readme';
import OptionList from './OptionList';

// tslint:disable-next-line no-var-requires
const readme = require('./README.md');

storiesOf('OptionList', module)
  .addDecorator(withReadme(readme))
  .add('default', () => (
    <OptionList
      definitions={[
        {
          key: text('key', 'cancellation_reason'),
          required: boolean('required', false),
          type: text('type', 'string'),
          description: [
            'The reason given for the ',
            <a href="#" key="link">
              cancellation
            </a>,
            ' by the member support employee.',
          ],
          default: text('default value', 'null'),
          values: array('possible values', [
            'duplicate',
            'fraudulent',
            'requested_by_customer',
          ]),
        },
      ]}
    />
  ))
  .add('multiple options', () => (
    <OptionList
      definitions={[
        {
          key: 'product_returned',
          required: false,
          type: 'boolean',
          description:
            'Whether or not the product has been returned by the customer.',
          default: 'false',
        },
        {
          key: 'status',
          required: true,
          type: 'string',
          description: [
            'Determines the current ',
            <a href="#" key="link">
              activity status
            </a>,
            ' of the account holder.',
          ],
          default: 'active',
          values: ['active', 'inactive'],
        },
      ]}
    />
  ));
