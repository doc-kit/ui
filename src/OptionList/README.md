# Option List

Useful to render options, parameters, or arguments in a stylish list structure.

## Features

- Automatically builds the sentence format for the option based on the props passed into the component.
- Accepts pattern to allow HTML such as links to be built into the description sentence.

## Usage

**Import**:

`import { OptionList } from '@doc-kits/react';`

**JSX**:

```js
<OptionList
  tableTitle="Arguments",
  definitions={[
    {
      key: "cancellation_reason",
      required: true,
      type: "string",
      description: "The reason given for the cancellation."
      default: "null",
      values: [
        "duplicate",
        "fraudulent",
        "requested_by_customer",
      ],
    },
    {...},
  ]}
/>
```

## Props

| Property               | Type                                | Required | Default | Description                              |
| ---------------------- | ----------------------------------- | -------- | ------- | ---------------------------------------- |
| tableTitle             | string                              | no       | Options | Sets the title for the table             |
| definitions            | Array[Objects]                      | yes      | null    | A list of option properties              |
| definition.key         | string                              | yes      | null    | A name for the key of the option         |
| definition.required    | boolean                             | no       | false   | Set true if this option is required      |
| definition.type        | string                              | yes      | null    | The type for the option                  |
| definition.description | string, Array[string / HTMLElement] | no       | null    | A description for the option             |
| definition.default     | string                              | no       | null    | The default value of the option          |
| definition.values      | Array[string]                       | no       | null    | A list of possible values for the option |

## Styles

The default styles injected into `OptionList` are shown below.

```js
const styles = {
  wrapper: {
    borderBottom: '1px solid #cbd8de',
    color: '#4a6271',
    fontSize: ['0.875rem', '1rem'],
    lineHeight: '1.5',
  },
  title: {
    color: '#4a6271',
    fontSize: 'inherit',
    borderBottom: '1px solid #cbd8de',
    paddingBottom: '0.313em',
    textTransform: 'uppercase',
  },
  option: {
    borderTop: '1px solid #edf3f3',
    padding: '1.143em 0',
  },
  key: {
    color: '#334049',
    fontSize: 'inherit',
    fontWeight: '500',
  },
  optional: {
    color: '#939da3',
    fontSize: '1em',
  },
  required: {
    color: '#ffae54',
    fontSize: '0.857em',
    border: '1px solid #ffae54',
    borderRadius: '10px',
    padding: '0 0.571em',
    textTransform: 'uppercase',
  },
  type: {
    color: '#939da3',
    fontSize: 'inherit',
  },
  description: {
    color: 'inherit',
    fontSize: 'inherit',
    margin: '1em 0 0 0',
  },
  values: {
    color: '#939da3',
  },
  code: {
    backgroundColor: '#fafcfc',
    borderRadius: '3px',
    border: `1px solid ${darken(0.05, '#fafcfc')}`,
    color: '#b93d6a',
    padding: '0 0.4em',
    fontFamily: 'inherit',
  },
  palette: {
    link: '#0099e5',
  },
};
```

## Notes

- You will _almost always_ want to add a punctuation period at the end of your `description` as it will be entered as a sentence.

- You may want to add a link into the description. To do so, you must use a mixed Array of strings and an `a` HTMLElement. It would look like the following: <br><br>

  ```js
  {
    description: [
      'The reason given for the ',
      <a href="#" key="link">cancellation</a>,
      ' by the member support employee.',
    ],
  }
  ```
