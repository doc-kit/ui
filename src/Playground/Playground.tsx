import styled from '@emotion/styled';
import React, { Component, ReactNode } from 'react';
import posed from 'react-pose';
import { ThemeProvider } from 'emotion-theming';
import { LiveProvider, LiveError, LivePreview } from 'react-live';
import Resizable from 're-resizable';
import reactElementToJSXString from 'react-element-to-jsx-string';
import merge from 'deepmerge';
import Editor from './components/Editor';
import Handle from './components/Handle';
import { Copy, Code, Refresh, Source } from './components/Actions';
import constructTheme from '../toolkit/constructTheme';
import defaultStyles from './styles';

interface Props {
  children: ReactNode;
  components?: object;
  mode?: string;
  actions?: {
    copy?: boolean;
    code?: boolean;
    refresh?: boolean;
    source?: boolean;
  };
  source?: string;
  editor?: object;
  readonly withStyles?: object;
  styles?: object;
  [propName: string]: any;
}

interface State {
  code: string;
  toggles: {
    showing: boolean;
    copying: boolean;
    refreshing: boolean;
  };
}

interface ToggleOptions {
  withTimeout: boolean;
}

interface WrapperProps {
  code: string;
  mode: string;
  scope?: object;
}

interface PreviewProps {
  align: string;
}

const Wrapper = styled(LiveProvider)<WrapperProps>`
  ${p => p.theme.mq(p.theme.wrapper)};

  border: ${p => p.mode === 'containerOnly' && 'none'};
  width: ${p => p.mode === 'containerOnly' && '100%'};
`;

const StyledPreview = styled(LivePreview)<PreviewProps>`
  ${p => p.theme.mq(p.theme.preview)};

  justify-content: ${p => p.align};
`;

const StyledError = styled(LiveError)`
  ${p => p.theme.mq(p.theme.error)};
`;

const Actions = styled.div`
  ${p => p.theme.mq(p.theme.actions)};
`;

const EditorWrapper = posed.div({
  closed: { height: 0, overflow: 'hidden' },
  open: { height: 'auto' },
});

class Playground extends Component<Props, State> {
  private static defaultProps = {
    mode: 'full',
    align: 'flex-start',
    actions: {
      copy: true,
      code: true,
      refresh: true,
      source: true,
    },
  };

  public state = {
    code: reactElementToJSXString(this.props.children, {
      useFragmentShortSyntax: false,
      showDefaultProps: false,
    }),
    toggles: {
      showing: false,
      refreshing: false,
      copying: false,
    },
  };

  private resizableProps = {
    minWidth: 360,
    maxWidth: '100%',
    enable: {
      bottom: false,
      bottomLeft: false,
      bottomRight: false,
      left: false,
      right: this.props.mode !== 'containerOnly' && true,
      top: false,
      topLeft: false,
      topRight: false,
    },
    handleComponent: {
      right: () => this.props.mode !== 'containerOnly' && <Handle />,
    },
    handleStyles: {
      right: {
        right: 8,
      },
    },
  };

  public render() {
    const toggle = this.toggle;
    const { code, toggles } = this.state;
    const {
      components,
      mode,
      align,
      actions,
      editor,
      source,
      withStyles,
      styles,
      ...restOfProps
    } = merge(Playground.defaultProps, this.props);

    const disableActions = mode !== 'full' || (!actions.code && !source);
    const componentTheme = () =>
      constructTheme(defaultStyles, withStyles, styles);

    return (
      <ThemeProvider theme={componentTheme}>
        <Resizable {...this.resizableProps}>
          <Wrapper mode={mode} code={code} scope={components} {...restOfProps}>
            <StyledPreview align={align} />

            {!disableActions && (
              <Actions>
                {actions.copy && (
                  <Copy
                    code={code}
                    show={toggles.showing}
                    inProgress={toggles.copying}
                    toggle={toggle}
                  />
                )}

                {actions.code && <Code onClick={this.toggleCode} />}

                {actions.refresh && (
                  <Refresh inProgress={toggles.refreshing} toggle={toggle} />
                )}

                {source && actions.source && <Source link={source} />}
              </Actions>
            )}

            <StyledError />
            <EditorWrapper pose={toggles.showing ? 'open' : 'closed'}>
              <Editor
                code={code}
                options={editor}
                handleChange={this.handleChange}
              />
            </EditorWrapper>
          </Wrapper>
        </Resizable>
      </ThemeProvider>
    );
  }

  private handleChange = (code: string) => {
    this.setState({ code });
  };

  private toggleCode = () => {
    this.toggle('showing');
  };

  private toggle = (
    toggle: string,
    options: ToggleOptions = { withTimeout: false }
  ) => {
    const { toggles } = this.state;

    this.setState(
      prevState => ({
        toggles: { ...toggles, [toggle]: !prevState.toggles[toggle] },
      }),
      () => {
        if (options.withTimeout) {
          setTimeout(() => {
            this.setState(prevState => ({
              toggles: {
                ...prevState.toggles,
                [toggle]: !prevState.toggles[toggle],
              },
            }));
          }, 1000);
        }
      }
    );
  };
}

export default Playground;
