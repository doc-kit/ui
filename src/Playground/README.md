# Playground

Useful to render a component as an example. The Playground allows users to edit code and preview the changes live.

<p align="left">
  <img src="https://raw.githubusercontent.com/doc-kits/react/master/images/playground.png?raw=true" width="1000">
</p>

**[Demo](https://doc-kits.github.io/react/?selectedKind=Playground&selectedStory=default&full=0&addons=1&stories=1&panelRight=0&addonPanel=REACT_STORYBOOK%2Freadme%2Fpanel)**

## Features

- Live preview, code, and errors
- A resizable preview area to show responsiveness
- Set a `mode` to switch preview style
- Complete control over which actions are available
- Provide a link back to the original source code
- Set the preview alignment via Flexbox's `justify-content`
- Ability to render class and functional components as well as pure JSX

## Usage

**Import**:

`import { Playground } from '@doc-kits/react';`

**JSX**:

When passing a React component as children, the component must be passed to the `components` prop as context for the Playground.

```js
import { Button } from 'my-library';

<Playground components={{ Button }}>
  <Button>Click Me!</Button>
</Playground>
```

See [more examples](#examples) below.

## Props

| Property   | Type   | Required | Default    | Description                                                                                                                        |
| ---------- | ------ | -------- | ---------- | ---------------------------------------------------------------------------------------------------------------------------------- |
| components | object | yes*     | null       | The set of components the Playground will use as context for which components are being displayed                                  |
| mode       | string | no       | full       | A preset representation of the Playground. Possible values are: `full`, `handleOnly`, and `containerOnly`.                         |
| align      | string | no       | flex-start | Any valid value for Flexbox's `justify-content`                                                                                    |
| actions    | object | no       | true       | Set individual action icons to `false` to remove from the view. The possible actions are: `copy`, `code`, `refresh`, and `source`. |
| source     | string | no       | null       | An absolute URL to the original source code for the component being displayed in the Playground                                    |
| editor     | object | no       | null       | Any valid [CodeMirror options](https://codemirror.net/doc/manual.html#config) to be set on the code editor                         |

*Only required when passing a React component as `children`. If passing JSX or HTML directly, `components` is _not_ required.

## Styles

The default styles injected into `Playground` are shown below.

```js
const styles = {
  wrapper: {
    borderRadius: '3px',
    borderBottomRightRadius: 0,
    borderTopRightRadius: 0,
    border: '1px solid #d8d8d8',
    width: 'calc(100% - 18px)',
  },
  preview: {
    display: 'flex',
    flexWrap: 'wrap',
    padding: '0.75em',
    '& > *': {
      margin: '0.5em',
    },
  },
  error: {
    backgroundColor: '#f8d7da',
    color: '#721c24',
    fontSize: '0.875rem',
    padding: '0.25em 1em',
  },
  editor: {
    backgroundColor: '#fbfcfd',
    borderTop: '1px solid #d8d8d8',
    '& .CodeMirror': {
      backgroundColor: '#fbfcfd',
      fontFamily: "'Inconsolata', monospace",
      fontSize: '0.875em',
      height: 'auto',
    },
    '& .CodeMirror-lines': {
      padding: '1.5em 0',
    },
  },
  actions: {
    alignItems: 'center',
    backgroundColor: '#f7fbfa',
    display: 'flex',
    fontSize: ['0.625rem', '0.625rem', '0.688rem', '0.688rem', '0.750rem'],
    justifyContent: 'flex-end',
    paddingTop: '0.555em',
    paddingBottom: '0.355em',
  },
  action: {
    alignItems: 'center',
    animation: 'fadeIn 0.25s',
    appearance: 'none',
    background: 'none',
    border: 'none',
    color: '#a7b0ca',
    display: 'flex',
    lineHeight: 1,
    marginRight: '1em',
    opacity: 1,
    outline: 'none',
    padding: 0,
    '&:hover': {
      color: 'darken(0.1, #a7b0ca)',
      cursor: 'pointer',
    },
    '& a': {
      alignItems: 'center',
      color: 'inherit',
      display: 'flex',
      textDecoration: 'none',
    },
  },
  actionText: {
    marginLeft: '0.35em',
    textTransform: 'uppercase',
  },
  handle: {
    alignItems: 'center',
    backgroundColor: '#f9f7f3',
    borderBottomRightRadius: '3px',
    borderTopRightRadius: '3px',
    border: '1px solid #d8d8d8',
    boxSizing: 'border-box',
    display: 'flex',
    flexDirection: 'column',
    height: '100%',
    justifyContent: 'center',
    width: '18px',
  },
  handleIcon: {
    opacity: '0.15',
    '&:last-child': {
      marginTop: '-0.250em',
    },
  },
};
```

## Examples

**Configuration**

This example shows how each of the available `props` might be configured to your desired state.

```js
import { Button } from 'my-library';

<Playground
  components={{ Button }}
  mode="containerOnly"
  align="center"
  actions={{ copy: false, refresh: false }}
  source="https://www.github.com/repo/button.js"
  editor={{ lineNumbers: true }}
>
  <Button>Click Me!</Button>
</Playground>
```

**Multiple Components**

When rendering multiple components into the Playground you must wrap them in a `<Fragment>` or `<div>` as you would need to in normal JSX.

```js
import { Fragment } from 'React';

<Playground>
  <Fragment>
    <Button>Copy</Button>
    <Button>Paste</Button>
    <Button>Clear</Button>
  </Fragment>
</Playground>
```

**Pure JSX**

If rendering simple JSX you do not need to pass any `props` to the Playground.

```js
<Playground>
  <div>Hello, World!</div>
</Playground>
```

## Notes

- The code editor that is part of the Playground uses a `font-family` of `Inconsolata` in the Storybook examples. If you would like to use the same font you must [inject](https://fonts.google.com/specimen/Inconsolata) `Inconsolata` into your application. Otherwise, you are free to override and use which ever font you desire for the editor.
- The Playground uses an open-source library called `react-live` under the hood. You can see more styles of JSX code that can be rendered into the Playground on their [demo site](https://react-live.kitten.sh/).
