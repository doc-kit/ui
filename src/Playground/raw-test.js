<ColorBox name="Red Barn" hex="#F00000" />;
