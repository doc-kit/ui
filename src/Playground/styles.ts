import { keyframes } from '@emotion/core';
import { darken } from 'polished';
import { base, palette } from '../toolkit/theme';

const HANDLE_SIZE = '18px';
const ICON_SIZE = '1.15em';

const colors = {
  editor: '#fbfcfd',
  actions: '#f7fbfa',
  handle: '#f9f7f3',
  icon: '#a7b0ca',
  error: {
    background: '#f8d7da',
    text: '#721c24',
  },
};

const fadeIn = keyframes`
  0% {
		opacity: 0;
	}
	100% {
		opacity: 1;
	}
`;

const styles = {
  wrapper: {
    ...base,
    borderRadius: '3px',
    borderBottomRightRadius: 0,
    borderTopRightRadius: 0,
    border: `1px solid ${palette.grayScale['200']}`,
    width: `calc(100% - ${HANDLE_SIZE})`,
  },
  preview: {
    display: 'flex',
    flexWrap: 'wrap',
    padding: '0.75em',
    '& > *': {
      margin: '0.5em',
    },
  },
  error: {
    backgroundColor: colors.error.background,
    color: colors.error.text,
    fontSize: '0.875rem',
    padding: '0.25em 1em',
  },
  editor: {
    backgroundColor: colors.editor,
    borderTop: `1px solid ${palette.grayScale['200']}`,
    '& .CodeMirror': {
      backgroundColor: colors.editor,
      fontFamily: 'monospace',
      fontSize: '0.875em',
      height: 'auto',
    },
    '& .CodeMirror-lines': {
      padding: '1.5em 0',
    },
  },
  actions: {
    alignItems: 'center',
    backgroundColor: colors.actions,
    display: 'flex',
    justifyContent: 'flex-end',
    paddingTop: '0.925em',
    paddingBottom: '0.725em',
  },
  action: {
    alignItems: 'center',
    animation: `${fadeIn} 0.25s`,
    appearance: 'none',
    background: 'none',
    border: 'none',
    color: colors.icon,
    display: 'flex',
    fontSize: ['0.625rem', '0.625rem', '0.688rem', '0.688rem', '0.750rem'],
    lineHeight: 1,
    marginRight: '1.75em',
    opacity: 1,
    outline: 'none',
    padding: 0,
    '&:hover': {
      color: `${darken(0.1, colors.icon)}`,
      cursor: 'pointer',
    },
    '& a': {
      alignItems: 'center',
      color: 'inherit',
      display: 'flex',
      textDecoration: 'none',
    },
  },
  actionText: {
    marginLeft: '0.35em',
    textTransform: 'uppercase',
  },
  handle: {
    alignItems: 'center',
    backgroundColor: colors.handle,
    borderBottomRightRadius: '3px',
    borderTopRightRadius: '3px',
    border: `1px solid ${palette.grayScale['200']}`,
    boxSizing: 'border-box',
    display: 'flex',
    flexDirection: 'column',
    height: '100%',
    justifyContent: 'center',
    width: `${HANDLE_SIZE}`,
  },
  handleIcon: {
    opacity: '0.15',
    '&:last-child': {
      marginTop: '-0.250em',
    },
  },
};

export default styles;
export { ICON_SIZE };
