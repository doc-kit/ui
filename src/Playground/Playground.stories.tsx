import { text, boolean, select } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React, { Fragment } from 'react';
import { withReadme } from 'storybook-readme';
import Playground from './Playground';
import ColorBox from '../ColorBox';
import OptionList from '../OptionList';

// tslint:disable-next-line no-var-requires
const readme = require('./README.md');

storiesOf('Playground', module)
  .addDecorator(withReadme(readme))
  .add('default', () => (
    <Playground
      components={{ OptionList }}
      mode={select(
        'mode',
        {
          full: 'full',
          handleOnly: 'handleOnly',
          containerOnly: 'containerOnly',
        },
        'full'
      )}
      source={text('source', 'http://www.mysite.com')}
      align={text('align', 'flex-start')}
      actions={{
        copy: boolean('copy icon', true),
        code: boolean('code icon', true),
        refresh: boolean('refresh icon', true),
        source: boolean('source icon', true),
      }}
    >
      <OptionList
        definitions={[
          {
            key: 'cancellation_reason',
            required: false,
            type: 'string',
            description: [
              'The reason given for the ',
              <a href="#" key="link">
                cancellation
              </a>,
              ' by the member support employee.',
            ],
            default: 'null',
            values: ['duplicate', 'fraudulent', 'requested_by_customer'],
          },
        ]}
      />
    </Playground>
  ))
  .add('multiple components', () => (
    <Playground
      components={{ ColorBox }}
      mode={select(
        'mode',
        {
          full: 'full',
          handleOnly: 'handleOnly',
          containerOnly: 'containerOnly',
        },
        'full'
      )}
      align={text('align', 'flex-start')}
      actions={{
        copy: boolean('copy icon', true),
        code: boolean('code icon', true),
        refresh: boolean('refresh icon', true),
      }}
      editor={{ lineNumbers: true }}
    >
      <Fragment>
        <ColorBox name="Red Barn" hex="#F00000" />
        <ColorBox name="Cream Popsicle" hex="#F09320" />
        <ColorBox name="Sky Blue" hex="#00AAE5" />
        <ColorBox name="Lavender" hex="#977BB6" />
      </Fragment>
    </Playground>
  ));
