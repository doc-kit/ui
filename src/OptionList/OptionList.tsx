/** @jsx jsx */
import { jsx } from '@emotion/core';
import styled from '@emotion/styled';
import * as merge from 'deepmerge';
import { ThemeProvider } from 'emotion-theming';
import { mq } from '../toolkit/responsive';
import settings from '../toolkit/settings';
import styles from './styles';

interface Definition {
  key: string;
  required?: boolean;
  type: string;
  description?: string | Array<string | any>;
  default?: string;
  values?: string[];
}

interface Props {
  tableTitle?: string;
  definitions: Definition[];
  localStyles?: any;
}

interface RequiredProps {
  required?: boolean;
}

const Wrapper = styled.div`
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;

  ${p => p.theme.mq(p.theme.wrapper)};

  a {
    color: ${p => p.theme.colors.link};
    text-decoration: none;

    &:hover {
      text-decoration: underline;
    }

    &:active,
    &:visited {
      color: ${p => p.theme.colors.link};
    }
  }
`;

const Title = styled.div`
  ${p => p.theme.mq(p.theme.title)};
`;

const Option = styled.div`
  ${p => p.theme.mq(p.theme.option)};

  &:first-of-type {
    border-top: none;
  }
`;

const Info = styled.div`
  align-items: baseline;
  display: flex;

  div {
    margin-right: 0.357em;

    &:last-of-type {
      margin-right: 0;
    }
  }
`;

const Key = styled.div`
  ${p => p.theme.mq(p.theme.key)};
`;

const Required = styled.div<RequiredProps>`
  border: ${p => (p.required ? p.theme.required.border : 'none')};
  border-radius: ${p => p.theme.required.borderRadius};
  color: ${p => (p.required ? p.theme.required.color : p.theme.optional.color)};
  font-size: ${p =>
    p.required ? p.theme.required.fontSize : p.theme.optional.fontSize};
  padding: ${p => (p.required ? p.theme.required.padding : '0')};
  text-transform: ${p =>
    p.required ? p.theme.required.textTransform : 'none'};
`;

const Type = styled.div`
  ${p => p.theme.mq(p.theme.type)};
`;

const Description = styled.p`
  ${p => p.theme.mq(p.theme.description)};
`;

const Values = styled.span`
  ${p => p.theme.mq(p.theme.values)};

  span:not(:first-of-type)::before {
    content: ', ';
  }

  span:nth-last-of-type(2):not(:first-of-type)::after {
    content: ',';
  }

  span:last-child:not(:only-of-type)::before {
    content: ' or ';
  }

  div:nth-of-type(1) {
    margin-bottom: 0.25em;
  }
`;

const Code = styled('code')`
  ${p => p.theme.mq(p.theme.code)};
`;

const OptionList = ({ tableTitle, definitions, localStyles }: Props) => {
  const sortedDefinitions = definitions.sort(
    (a: any, b: any) => b.required - a.required
  );
  const componentStyles = (appTheme: object = {}) =>
    merge.all([settings, styles, { mq }, appTheme, localStyles]);

  return (
    <ThemeProvider theme={componentStyles}>
      <Wrapper>
        <Title>{tableTitle}</Title>

        {sortedDefinitions.map(argument => {
          const requiredText = argument.required ? 'required' : 'optional';

          return (
            <Option key={argument.key}>
              <Info>
                <Key>{argument.key}</Key>
                <Required required={argument.required}>{requiredText}</Required>
                <Type>{argument.type}</Type>
              </Info>
              <Description>
                <span>{argument.description} </span>
                {argument.default && (
                  <span>
                    The default value is <Code>{argument.default}</Code>.
                  </span>
                )}
                {argument.values && <span> Possible values are </span>}
                {argument.values && (
                  <Values>
                    {argument.values.map((value: string, index: number) => (
                      <span key={index.toString()}>
                        <Code>{value}</Code>
                      </span>
                    ))}
                    .
                  </Values>
                )}
              </Description>
            </Option>
          );
        })}
      </Wrapper>
    </ThemeProvider>
  );
};

OptionList.defaultProps = {
  tableTitle: 'Options',
  localStyles: {},
  definitions: {
    required: false,
  },
};

export default OptionList;
