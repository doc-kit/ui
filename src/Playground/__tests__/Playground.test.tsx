import React, { Fragment } from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import OptionList from '../../OptionList';
import Playground from '../Playground';

describe('<Playground />', () => {
  it('should render with minimum requirements', () => {
    const tree = renderer
      .create(
        <Playground>
          <div>Playground</div>
        </Playground>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render with all available passed in props', () => {
    const tree = renderer
      .create(
        <Playground
          components={{ OptionList }}
          mode="handleOnly"
          actions={{ copy: false }}
          editor={{ lineNumbers: true }}
          source="https://jestjs.io"
          styles={{ error: { color: 'green' } }}
        >
          <div>Playground</div>
        </Playground>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should set state.code to a string received from children jsx', () => {
    const wrapper = shallow(
      <Playground>
        <Fragment>
          <div>Playground</div>
        </Fragment>
      </Playground>
    );

    expect(typeof wrapper.state('code')).toEqual('string');
  });

  it('should update state.code when handleChange called', () => {
    const wrapper = shallow(
      <Playground>
        <div>Playground</div>
      </Playground>
    );

    (wrapper.instance() as any).handleChange('<div>Different</div>');
    expect(wrapper.state('code')).toEqual('<div>Different</div>');
  });

  it('should set state.toggles.showing when toggleCode called', () => {
    const wrapper = shallow(
      <Playground>
        <div>Playground</div>
      </Playground>
    );

    (wrapper.instance() as any).toggleCode();
    expect((wrapper.state() as any).toggles.showing).toBeTruthy();
  });

  it('should setState for specified UI toggle when toggle method called', () => {
    const wrapper = shallow(
      <Playground>
        <div>Playground</div>
      </Playground>
    );

    (wrapper.instance() as any).toggle('refreshing');
    expect((wrapper.state() as any).toggles.refreshing).toBeTruthy();
    expect((wrapper.state() as any).toggles.showing).toBeFalsy();
    expect((wrapper.state() as any).toggles.copying).toBeFalsy();
  });

  it('should not render specific actions if turned off', () => {
    const wrapper = shallow(
      <Playground actions={{ copy: false, refresh: false }}>
        <div>Playground</div>
      </Playground>
    );

    expect(wrapper.find('Copy')).toHaveLength(0);
    expect(wrapper.find('Refresh')).toHaveLength(0);
    expect(wrapper.find('Code')).toHaveLength(1);
  });

  it('should render a source action if source passed', () => {
    const wrapper = shallow(
      <Playground source="source">
        <div>Playground</div>
      </Playground>
    );

    expect(wrapper.find('Source')).toHaveLength(1);
  });

  it('should not render a source action if source passed but action turned off', () => {
    const wrapper = shallow(
      <Playground source="source" actions={{ source: false }}>
        <div>Playground</div>
      </Playground>
    );

    expect(wrapper.find('Source')).toHaveLength(0);
  });

  it('should not render any actions if mode does not equal "full"', () => {
    const wrapper = shallow(
      <Playground mode="handleOnly">
        <div>Playground</div>
      </Playground>
    );

    expect(wrapper.find('Copy')).toHaveLength(0);
    expect(wrapper.find('Code')).toHaveLength(0);
    expect(wrapper.find('Refresh')).toHaveLength(0);
    expect(wrapper.find('Source')).toHaveLength(0);
  });

  it('should not render a Handle if mode is containerOnly', () => {
    const wrapper = shallow(
      <Playground mode="containerOnly">
        <div>Playground</div>
      </Playground>
    );

    expect(
      (wrapper.find('Resizable') as any).props().handleComponent.right()
    ).toBeFalsy();
  });
});
