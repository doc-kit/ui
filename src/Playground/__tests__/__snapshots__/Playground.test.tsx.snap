// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Playground /> should render with all available passed in props 1`] = `
<Resizable
  enable={
    Object {
      "bottom": false,
      "bottomLeft": false,
      "bottomRight": false,
      "left": false,
      "right": true,
      "top": false,
      "topLeft": false,
      "topRight": false,
    }
  }
  handleComponent={
    Object {
      "right": [Function],
    }
  }
  handleStyles={
    Object {
      "right": Object {
        "right": 8,
      },
    }
  }
  maxWidth="100%"
  minWidth={320}
>
  <LiveProvider
    className="css-5aoufr"
    code="<div>
  Playground
</div>"
    mode="handleOnly"
    scope={
      Object {
        "OptionList": [Function],
      }
    }
  >
    <LivePreview
      align="flex-start"
      className="css-xlsnbg"
    />
    <LiveError
      className="css-1mfynrb"
    />
    <div
      pose="closed"
    >
      <BaseEditor
        className="css-1j22wvk"
        onBeforeChange={[Function]}
        options={
          Object {
            "lineNumbers": true,
            "lineWrapping": true,
            "mode": "jsx",
            "scrollbarStyle": null,
            "tabSize": 2,
            "viewportMargin": Infinity,
          }
        }
        value="<div>
  Playground
</div>"
      />
    </div>
  </LiveProvider>
</Resizable>
`;

exports[`<Playground /> should render with minimum requirements 1`] = `
<Resizable
  enable={
    Object {
      "bottom": false,
      "bottomLeft": false,
      "bottomRight": false,
      "left": false,
      "right": true,
      "top": false,
      "topLeft": false,
      "topRight": false,
    }
  }
  handleComponent={
    Object {
      "right": [Function],
    }
  }
  handleStyles={
    Object {
      "right": Object {
        "right": 8,
      },
    }
  }
  maxWidth="100%"
  minWidth={320}
>
  <LiveProvider
    className="css-5aoufr"
    code="<div>
  Playground
</div>"
    mode="full"
  >
    <LivePreview
      align="flex-start"
      className="css-xlsnbg"
    />
    <div
      className="css-1aa4jxg"
    >
      <button
        className="css-o69cv9"
        onClick={[Function]}
      >
        <svg
          fill="none"
          height="1.15em"
          size="1.15em"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1.15em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="16 18 22 12 16 6"
          />
          <polyline
            points="8 6 2 12 8 18"
          />
        </svg>
        <div
          className="css-ivhnjk"
        >
          Code
        </div>
      </button>
      <button
        className="css-o69cv9"
        disabled={false}
        onClick={[Function]}
      >
        <div
          pose="stop"
          rotation={0}
        >
          <svg
            fill="none"
            height="0.9em"
            size="0.9em"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="0.9em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="23 4 23 10 17 10"
            />
            <polyline
              points="1 20 1 14 7 14"
            />
            <path
              d="M3.51 9a9 9 0 0 1 14.85-3.36L23 10M1 14l4.64 4.36A9 9 0 0 0 20.49 15"
            />
          </svg>
        </div>
        <div
          className="css-ivhnjk"
        >
          Refresh
        </div>
      </button>
    </div>
    <LiveError
      className="css-s7cjye"
    />
    <div
      pose="closed"
    >
      <BaseEditor
        className="css-umkx8t"
        onBeforeChange={[Function]}
        options={
          Object {
            "lineNumbers": false,
            "lineWrapping": true,
            "mode": "jsx",
            "scrollbarStyle": null,
            "tabSize": 2,
            "viewportMargin": Infinity,
          }
        }
        value="<div>
  Playground
</div>"
      />
    </div>
  </LiveProvider>
</Resizable>
`;
